### ISSUE TITLE
execvp() trouble in the git version

### ISSUE THREAD

**Issue by misutoneko** [uobikiemukot/yaft#54]
Hi,

First of all, thanks for this great little program, it's awesome.

I think I found a bug in the git version (the 0.2.9 version is not affected).
I've tested this in qemu with FreeBSD and NetBSD guests (i386, no X, using the local fb console).
I'm not sure if this problem affects other platforms, but it probably does.
There was at least some Linux lockup issue which might be related.
In the BSD guests the symptom is that running "yaft" only briefly flashes the screen and nothing else happens.
Also, no 256 colors, printing out $TERM shows that it stayed the same etc.

I tested a few git versions and it seems this problem was introduced with commit b5904bb (dated 2018-04-29).
The commit adds some command line parsing and replaces a execl() call with a execvp() one.
Which is all fine and good, but in yaft.c (line 190 or so), the third argument to fork_and_exec() is "argv + 1" which
seems to be incorrect (should be just "argv" without the +1).

As to why, well, the man page of execvp() says:
"The first argument, by convention, should point to the filename associated with the file being executed."

But that "first argument" here refers to the argument list (which is given as the second argument to execvp()).
So it's easy to get confused there.

I'm a bit baffled why this hasn't been reported before, as this bug has been in git for a couple of years already.
Everyone just happily using 0.2.9? OK yaft is in a specific niche, but still..

**Comment by hackerb9** [comment]
@misutoneko Could this be a difference between BSD and Linux? I tried the change you suggested (removing `+ 1`) and it breaks yaft so that, when given a command to execute, it just flashes. (E.g., `yaft sleep 5`). It makes sense to me to use argv+1 as that is the first argument, skipping over the `yaft` command  in argv[0].

**Comment by misutoneko** [comment]
Hi! Thanks, that's interesting.
I think I did try this in Linux too, but not with those extra parameters like you did.
Was it working in 0.2.9?

What happens if you spell it all out, like this?
`yaft /bin/bash -c "sleep 5"`

Or you could try making a shell script out of it (something that just prints out $TERM and then calls sleep).

The origin of the confusion is that the treatment of argv[] depends on what exact exec* function is being used:
If there's p in the function name, argv[0] is not used to locate the executable.
So argv[0] is kind of useless with the p variants, but it's there if anyone wants it.
The implementation details can differ, but I'd like to trust the man page on the actual usage.

**Comment by hackerb9** [comment]
I think I figured out where the confusion came from. The "argv" being referred to is yaft's argv, not the argv that is being built for the exec. Adding this fragment before the fork_and_exec may help:

```C
	char **p=argv; int t=0;
	while (p && *p) {
	  fprintf(stderr, "argv[%d] is '%s'\t", t++, *p++);
	}
```

For me, that gives the output:

    argv[0] is './yaft'     argv[1] is '/bin/bash'     argv[2] is '-c'     argv[3] is 'sleep 5'

Perhaps I am misunderstanding something, but I believe the argv+1 is correct. Now the mystery is why removing `+1` makes it work for you in BSD. Could you please double-check your results? Thanks!

**Comment by misutoneko** [comment]
Yes, double-checking is good, it has helped me before...
I even have my old vm from 3+ years ago(!), so I took another look.
So... I don't know how I missed it the first time around, but the default shell depends on OS.
In NetBSD it's **/bin/csh**.

Now if I launch yaft with ./yaft in NetBSD, csh doesn't like it for some reason and I get a coredump about it.
However, if I edit conf.h to use /usr/pkg/bin/bash instead, it works the same way as in Linux (=no crash).
The same crash can be reproduced in Linux too, if you change yaft's default shell to csh.
To be clear, by yaft's default shell I mean the one in the source file conf.h.

EDIT: If I use an imaginary shell such as /bin/shellfish, there's a flash but no crash.

**Comment by hackerb9** [comment]
Have you tried examining the corefile with `gdb ./yaft yaft.core`? Does the backtrace (`bt`) give any clues?

What happens if you change it to a shell that doesn't exist at all, like /bin/shellfish? Does it still coredump?

_____

_Sidenote: I think NetBSD switched their default user shell to the Almquist shell (ash). Not that it matters: yaft shouldn't core dump._

**Comment by misutoneko** [comment]
The gdb backtrace wasn't all that helpful because there are no symbols, but I suppose the crash happens in [s_strcmp().](https://github.com/NetBSD/src/blob/trunk/bin/csh/str.c#L260)
The call is located right at the beginning of main(), via some macros:
https://github.com/NetBSD/src/blob/trunk/bin/csh/csh.c#L235

Linux also generated some nice logs about it, here's the relevant part:
Mar  2 11:31:41 sparkyhost kernel: [ 1463.001606] csh[1794]: segfault at 0 ip 0041c6ea sp bffff3c4 error 4 in bsd-csh[402000+1c000]
Mar  2 11:31:41 sparkyhost kernel: [ 1463.001611] Code: 8b 54 24 04 31 c0 66 83 3a 00 75 04 c3 8d 76 00 83 c0 01 66 83 3c 42 00 75 f6 c3 8d 74 26 00 90 56 53 8b 44 24 0c 8b 5c 24 10 **<0f>** b7 08 0f bf 13 66 85 c9 74 1b 66 39 d1 74 36 b8 01 00 00 00 66

This is what it looks like in the Linux binary:

$ objdump -d /bin/bsd-csh |grep -C6 0f.bf.13

    1c6df:	90                   	nop
    1c6e0:	56                   	push   %esi
    1c6e1:	53                   	push   %ebx
    1c6e2:	8b 44 24 0c          	mov    0xc(%esp),%eax
    1c6e6:	8b 5c 24 10          	mov    0x10(%esp),%ebx
    1c6ea:	0f b7 08             	movzwl (%eax),%ecx
    1c6ed:	0f bf 13             	movswl (%ebx),%edx
    1c6f0:	66 85 c9             	test   %cx,%cx
    1c6f3:	74 1b                	je     1c710 <__cxa_finalize@plt+0x1a080>
    1c6f5:	66 39 d1             	cmp    %dx,%cx
    1c6f8:	74 36                	je     1c730 <__cxa_finalize@plt+0x1a0a0>
    1c6fa:	b8 01 00 00 00       	mov    $0x1,%eax
    1c6ff:	66 85 d2             	test   %dx,%dx

My interpretation: csh never anticipated the situation that argv could ever be NULL.
(So this is also a csh bug that we just exposed here?)
Since the command line I used to launch yaft was just ./yaft, the +1 causes argv to be nonexistant (argc is 0).

EDIT: So to take care of both possibilities, the argument needs to be something like (argc<2)?argv:argv+1.

**Comment by hackerb9** [comment]
Okay, I tried using the version of `csh` that comes with Debian GNU/Linux and it works fine for me. No crash, no coredump. This must be a bug in bsd-csh. I suggest compiling it from source so you can track down the problem and report it to NetBSD.

**Comment by misutoneko** [comment]
That's interesting, And you launched it with just ./yaft?
I was using a Debianish distro (Sparky) too, but it's an old version and there could differences to vanilla Debian.
(Debian does have its own patchset... Idk but I'd be tempted to think any such bugfixes would travel upstream...)

Seems I can't do full testing (or report for that matter) but copying s_strcmp() from str.c (plus some #defines) and calling the function with NULL as the first parameter results in SIGSEGV.
So that seems quite obvious. Ofc there's a bigger philosophical question about NULL testing in general.
In normal circumstances this situation doesn't happen at all... so all the NULL testing could be considered useless bloat?
Thinking about these kind of things seems like a good way to induce a headache...

**Comment by hackerb9** [comment]
> That's interesting, And you launched it with just ./yaft?

Correct. I edited the conf file you mentioned. I understand what you mean about a headache. 

If Sparky is derived from Debian, then you should be able to get the source code for any package easily. For example,

```
apt source csh
```
When I do that, I see that while there are some patches Debian adds, none of them seem to be about the problem you're seeing. It may be a different version of csh. My Debian is unpacking something called csh_20110502.orig.tar.gz, which sounds pretty old. 

I also notice that the program you're running is called "bsd-csh", but no such package is available from Debian.
